### ISSUE TITLE
Attempt for porting to OSX

### ISSUE THREAD

**Issue by zbyna** [dliw/fpCEF3#12]
Hi,
I am trying to adopt fpCEF3 for osx (Carbon widgets). I am able to load libraries but then app crashed.
Any pointing me is welcomed.
    Thanks
            zbyna
**Lazarus Call Stack:**

```
#0 ?? at :0
#1 cef_zip_reader_create at :0
#2 cef_zip_reader_create at :0
#3 cef_zip_reader_create at :0
#4 cef_browser_host_create_browser_sync at :0
#5 CEF3LIB_$$_CEFBROWSERHOSTCREATEBROWSERSYNC$crc4DE1F0D0 at :0
#6 CEF3LCL$_$TCUSTOMCHROMIUM_$__$$_CREATEBROWSER at :0
#7 CEF3LCL$_$TCUSTOMCHROMIUM_$__$$_CREATEWND at :0
#8 CONTROLS$_$TWINCONTROL_$__$$_CREATEWND at :0
#9 CONTROLS$_$TWINCONTROL_$__$$_HANDLENEEDED at :0
#10 CONTROLS$_$TWINCONTROL_$__$$_CREATEWND at :0
#11 FORMS$_$TSCROLLINGWINCONTROL_$__$$_CREATEWND at :0
#12 FORMS$_$TCUSTOMFORM_$__$$_CREATEWND at :0
#13 FORMS$_$TFORM_$__$$_CREATEWND at :0
#14 CONTROLS$_$TWINCONTROL_$__$$_CREATEHANDLE at :0
#15 CONTROLS$_$TWINCONTROL_$__$$_HANDLENEEDED at :0
#16 FORMS$_$TAPPLICATION_$__$$_CREATEFORM$TCOMPONENTCLASS$formal at :0
#17 PASCALMAIN at project1.lpr:18
#18 SYSTEM_$$_FPC_SYSTEMMAIN$LONGINT$PPCHAR$PPCHAR at :0
#19 _start at :0
#20 start at :0
```

**Mac system log:**

```
Crashed Thread:        0  CrBrowserMain  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000

VM Regions Near 0:
--> 
    __TEXT                 0000000000010000-00000000002ed000 [ 2932K] r-x/rwx SM=COW  /Users/USER/Documents/*/project1.app/Contents/MacOS/project1

Application Specific Information:
objc_msgSend() selector name: addSubview:


Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   libobjc.A.dylib                 0x9434b0ab objc_msgSend + 27
1   org.chromium.ContentShell.framework 0x07ab1c06 cef_zip_reader_create + 20710
2   org.chromium.ContentShell.framework 0x07ab1638 cef_zip_reader_create + 19224
3   org.chromium.ContentShell.framework 0x07ab13cd cef_zip_reader_create + 18605
4   org.chromium.ContentShell.framework 0x07a665a8 cef_browser_host_create_browser_sync + 952
5   com.company.project1            0x0028acdc CEF3LIB_$$_CEFBROWSERHOSTCREATEBROWSERSYNC$crc4DE1F0D0 + 220
6   com.company.project1            0x0028d729 CEF3LCL$_$TCUSTOMCHROMIUM_$__$$_CREATEBROWSER + 217
7   com.company.project1            0x0028d475 CEF3LCL$_$TCUSTOMCHROMIUM_$__$$_CREATEWND + 21
8   com.company.project1            0x001c4871 CONTROLS$_$TWINCONTROL_$__$$_HANDLENEEDED + 97
9   com.company.project1            0x001c3ead CONTROLS$_$TWINCONTROL_$__$$_CREATEWND + 2685
10  com.company.project1            0x0003a763 FORMS$_$TSCROLLINGWINCONTROL_$__$$_CREATEWND + 51
11  com.company.project1            0x00041557 FORMS$_$TCUSTOMFORM_$__$$_CREATEWND + 39
12  com.company.project1            0x00042714 FORMS$_$TFORM_$__$$_CREATEWND + 52
13  com.company.project1            0x001c3421 CONTROLS$_$TWINCONTROL_$__$$_CREATEHANDLE + 33
14  com.company.project1            0x001c4871 CONTROLS$_$TWINCONTROL_$__$$_HANDLENEEDED + 97
15  com.company.project1            0x0004ab55 FORMS$_$TAPPLICATION_$__$$_CREATEFORM$TCOMPONENTCLASS$formal + 261
16  com.company.project1            0x000116fb PASCALMAIN + 91
17  com.company.project1            0x00034605 FPC_SYSTEMMAIN + 69
18  com.company.project1            0x0001167d _start + 212
19  com.company.project1            0x000115a8 start + 40
```

**Console log:**

```
Last login: Mon Jun  1 22:32:39 on ttys000
ZBYNEK-MBP:~ zbyna$ /Users/zbyna/Documents/Code\ Typhon\ Pokusy/CEF3_pokus/project1.app/Contents/MacOS/project1 ; exit;
Timer created.
ClientHandler instances: 1
Warning: Using Carbon custom control instead of not implemented win control!
CefInitDefault
CefInitialize
CefLoadLibrary ... /Users/zbyna/Documents/Code Typhon Pokusy/CEF3_pokus/project1.app/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework
-> loaded
2015-06-01 22:36:38.051 project1[1515:55237] Internals of CFAllocator not known; out-of-memory failures via CFAllocator will not result in termination. http://crbug.com/45650
[0601/223638:WARNING:resource_bundle.cc(280)] locale_file_path.empty()
CefInitialize - zdá se proběhlo
CefBrowserHostCreateBrowserSync in cef3lib.pas
2015-06-01 22:36:38.163 project1 Helper[1516:55286] Internals of CFAllocator not known; out-of-memory failures via CFAllocator will not result in termination. http://crbug.com/45650
[0601/223638:WARNING:resource_bundle.cc(280)] locale_file_path.empty()
Segmentation fault: 11
logout
```

**Comment by dliw** [comment]
Hi, nice to see someone trying...
Unfortunately I have no experience with OS X at all, so I'm afraid I can't help much

I assume you already checked that
- you're using the right library version (3.1750)
- your OS X meets the minimum requirements (see the download page)
- the application bundle is set up correctly
- chrome-sandbox is working correctly (if it exists on Mac)

Searching for `Internals of CFAllocator ***` gives a lot of results, so this might be a hint for the crash. When working on the Linux side I had these crashes if certain calls happend on the wrong thread or pointers were invalid.
But without seeing your code it's hard to say anything more specific...

In general I would recommend to start with a minimal test case (e.g. by porting `cefsimple` to Pascal) and no LCL integration.
Afterwards it might be a good idea to try to get the OSR example to work, as the class abstraction is used with no LCL integration.

Btw. the call stack reminds me of the crashes I had on Linux, when trying to update to 3.2171 (Win version was/is working well)...
... I even could reproduce them with a minimal C program - so there is always the possibility that there is a bug in the library itself (more precisely: in the C abstraction of the library)...

Are you calling `cef_zip_reader_create` or where does this call come from?

**Comment by zbyna** [comment]
Hi,
thanks for response and tips.
> you're using the right library version (3.1750) 

yes
> your OS X meets the minimum requirements (see the download page)

yes
> the application bundle is set up correctly

   I believe yes, but it was not piece of cake job :smile: (new tools for me like otool, install_name_tool) 
> chrome-sandbox is working correctly (if it exists on Mac)

seems like it does not exist on Mac

> Internals of CFAllocator *** 

this message is presented when using official cefclient application, so I think it is not
source of bug with ```cef_zip_reader_create```

> Are you calling ```cef_zip_reader_create``` or where does this call come from?

I am not aware of changing something related this routine in package code and gdb on osx does not help too much (problems with settings breakpoints for some different projects with same settings),  but it is a good idea you pointing out **to start with simple case with no LCL integration**. I will try.

**Comment by dliw** [comment]
I just wondered why someone would call `cef_zip_reader_create` if everything else does not work - never mind :)

In general debugging CEF is not easy, I found it helpful to debug the renderer process separately (see the `Debugging` section in the readme), but I don't know, if this works on Mac the same way...

Don't hesitate to ask and good luck!

**Comment by zbyna** [comment]
So I started with simple app simillar to minimal example here in repository. Plugin works well as can be spotted from cef3 log, but I am ```not able to see output visually```. The same result  with lcl simple application. It seems I am stuck and time I can afford to play with  is over. I am affraid that issue is in fact that cef3 needs nsView pointer (Cocoa framework) and I am able obtain only HIViewRef pointers from Carbon framework. I am not able to judge.   If somebody wants to continue I may provide him with changes in source code (a few) I made.

**Comment by dliw** [comment]
I found this [doc](http://www.filibeto.org/unix/macos/lib/dev/documentation/Cocoa/Conceptual/CarbonCocoaDoc/CarbonCocoaIntegration.pdf), but information is little and I suspect there is no good way to embed a Cocoa wideget into Carbon...
... a way around would be to use OSR, redirect any mouse and key events and just draw the rendered image...

Would you mind sending me all the changes you made? I either put them directly into the repo or provide a link to a external download location, that way things don't get lost...

**Comment by zbyna** [comment]
Yes, this is very good source of information. I have learned all things about using Carbon I know sofar from this site docs.
I will send you all my changes, I have no problems with it. But I have no time to immediatelly react.
I enclose only console log, some items indicates that I somehow break apple security rules  see ```dyld: Library not loaded:```: ( I cut verbose items):
```
Last login: Tue Jun  9 12:45:25 on ttys000
ZBYNEK-MBP:~ zbyna$ /Users/zbyna/Documents/Code_Typhon_Pokusy/CEF3_pokus/mac_minimal/minimal.app/Contents/MacOS/minimal ; exit;
CefLoadLibrary ... /Users/zbyna/Documents/Code_Typhon_Pokusy/CEF3_pokus/mac_minimal/minimal.app/Contents/Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework
-> loaded
2015-06-09 13:10:39.073 minimal[1348:38420] Internals of CFAllocator not known; out-of-memory failures via CFAllocator will not result in termination. http://crbug.com/45650
[0609/131039:VERBOSE1:pref_proxy_config_tracker_impl.cc(148)] 0x83d010: set chrome proxy config service to 0x83e3a0
[0609/131039:VERBOSE1:pref_proxy_config_tracker_impl.cc(274)] 0x83d010: Done pushing proxy to UpdateProxyConfig
2015-06-09 13:10:39.134 minimal[1348:38420] WARNING: CFSTR("m\37777777705\37777777657jScrollView") has non-7 bit chars, interpreting using MacOS Roman encoding for now, but this will change. Please eliminate usages of non-7 bit chars (including escaped characters above \177 octal) in CFSTR().
2015-06-09 13:10:39.134 minimal[1348:38420] WARNING: CFSTR("m\37777777705\37777777657jVbox") has non-7 bit chars, interpreting using MacOS Roman encoding for now, but this will change. Please eliminate usages of non-7 bit chars (including escaped characters above \177 octal) in CFSTR().
2015-06-09 13:10:39.175 minimal[1348:38420] NSWindow warning: adding an unknown subview: <OpaqueView: 0xa27020>
2015-06-09 13:10:39.176 minimal[1348:38420] Call stack:
(
	0   AppKit                              0x9a373441 -[NSThemeFrame addSubview:] + 108
	1   Chromium Embedded Framework         0x06d0de47 _CTFontManagerUnregisterFontForData + 8646487
	2   Chromium Embedded Framework         0x0607c06d cef_time_delta + 327117
	3   Chromium Embedded Framework         0x05faec06 cef_zip_reader_create + 20710
	4   Chromium Embedded Framework         0x05fae638 cef_zip_reader_create + 19224
	5   Chromium Embedded Framework         0x05fae3cd cef_zip_reader_create + 18605
	6   Chromium Embedded Framework         0x05f635a8 cef_browser_host_create_browser_sync + 952
	7   minimal                             0x00011564 PASCALMAIN + 1556
	8   minimal                             0x00034625 FPC_SYSTEMMAIN + 69
	9   minimal                             0x00010e1d _start + 212
	10  minimal                             0x00010d48 start + 40
	11  ???                                 0x00000001 0x0 + 1
)
2015-06-09 13:10:39.216 minimal Helper[1349:38464] Internals of CFAllocator not known; out-of-memory failures via CFAllocator will not result in termination. http://crbug.com/45650
2015-06-09 13:10:39.296 minimal Helper[1350:38479] Internals of CFAllocator not known; out-of-memory failures via CFAllocator will not result in termination. http://crbug.com/45650
[0609/131039:ERROR:renderer_main.cc(226)] Running without renderer sandbox
[0609/131039:VERBOSE1:resource_loader.cc(202)] OnReceivedRedirect: http://www.google.com/
[0609/131040:VERBOSE1:resource_loader.cc(202)] OnReceivedRedirect: http://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA
[0609/131040:VERBOSE1:resource_loader.cc(304)] OnResponseStarted: https://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA&gws_rd=ssl
[0609/131040:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA&gws_rd=ssl" bytes_read = 1222
[0609/131040:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA&gws_rd=ssl" bytes_read = 3409
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA&gws_rd=ssl" bytes_read = 2964
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA&gws_rd=ssl" bytes_read = 0
[0609/131041:VERBOSE1:resource_loader.cc(608)] ResponseCompleted: https://www.google.cz/?gfe_rd=cr&ei=r8l2Vb2FMOqk8we_16mYBA&gws_rd=ssl
dyld: Library not loaded: @executable_path/../../../../Frameworks/Chromium Embedded Framework.framework/Chromium Embedded Framework
  Referenced from: /Users/zbyna/Documents/Code_Typhon_Pokusy/CEF3_pokus/mac_minimal/minimal.app/Contents/Frameworks/minimal Helper EH.app/Contents/MacOS/minimal Helper EH
  Reason: unsafe use of @executable_path in /Users/zbyna/Documents/Code_Typhon_Pokusy/CEF3_pokus/mac_minimal/minimal.app/Contents/Frameworks/minimal Helper EH.app/Contents/MacOS/minimal Helper EH with restricted binary
[0609/131041:VERBOSE1:resource_loader.cc(304)] OnResponseStarted: https://www.google.cz/images/hpp/logo-myaccount-hpp-48px.png
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/images/hpp/logo-myaccount-hpp-48px.png" bytes_read = 1921
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/images/hpp/logo-myaccount-hpp-48px.png" bytes_read = 0
[0609/131041:VERBOSE1:resource_loader.cc(608)] ResponseCompleted: https://www.google.cz/images/hpp/logo-myaccount-hpp-48px.png
[0609/131041:VERBOSE1:resource_loader.cc(304)] OnResponseStarted: https://ssl.gstatic.com/gb/images/i1_71651352.png
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://ssl.gstatic.com/gb/images/i1_71651352.png" bytes_read = 1000
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/images/srpr/logo11w.png" bytes_read = 658
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/images/srpr/logo11w.png" bytes_read = 0
[0609/131041:VERBOSE1:resource_loader.cc(608)] ResponseCompleted: https://www.google.cz/images/srpr/logo11w.png
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/images/nav_logo195.png" bytes_read = 1479
[0609/131041:VERBOSE1:resource_loader.cc(344)] OnReadCompleted: "https://www.google.cz/images/nav_logo195.png" bytes_read = 0
[0609/131041:VERBOSE1:resource_loader.cc(608)] ResponseCompleted: https://www.google.cz/images/nav_logo195.png
```

**Comment by zbyna** [comment]
Sorry for closing :smiley:

**Comment by dliw** [comment]
To me `[0609/131039:ERROR:renderer_main.cc(226)] Running without renderer sandbox` looks very much like the actual problem...

**Comment by zbyna** [comment]
I am sending files and diffs to actual master through mail.
As fo ``` Running without renderer ... ``` I thought that it is safe error : https://github.com/cefsharp/CefSharp/issues/501.

**Comment by zbyna** [comment]
Zdravíčko,

posílám ty soubory a upravený příklad minimal, který jsem používal na testování.

   Zbyňa

**Comment by dliw** [comment]
Yes, you are right, of course - that happens every time when running without the sandbox. I somehow forgot about it :(

**Comment by dliw** [comment]
The new version contains some of your changes - thank you!
For me it seems that there is no way to use CEF with the Carbon interface. Cocoa is the future anyway, but I don't know if this makes sense with the current state of Cocoa LCL.

**Comment by quartexNOR** [comment]
You cant use the same files on the mac. Download the CEF files compiled for OS X. Even though both are x86 and have the same symbols - there are differences. Also, what UI API you use doesnt really matter since the CEF library renders its own into a bytebuffer (which is the cool thing).
Every UI widgetset can display graphics.

**Comment by dliw** [comment]
> You cant use the same files on the mac
You mean the same files as on Linux? Well, that's more than obvious.

>Also, what UI API you use doesnt really matter since the CEF library renders its own into a bytebuffer
That's true when using CEF with OSR. For the default embed method - a LCL component - the UI API is crucial. If you read this issue more closely you will actually see why.

Mainly:
>there is no good way to embed a Cocoa widget into Carbon
CEF needs to be embedded into a Cocoa widget - although it draws everything itself and the LCL Cocoa interface isn't really ready AFAIK

On the other hand: QT on OS X might work already, but I don't have a Mac.

**Comment by skalogryz** [comment]
Hey Dliw.

There was a bounty recently posted to port fpcef3 to OSX
http://wiki.freepascal.org/Bounties#CEF3_Chromium_Embedded_On_MacOSX

I accepted it and succeeded. However, only Cocoa 64-bit is supported (due to CEF dropping 32-bit support for OSX). I forked your repository and you can found all the code (changes) here: https://github.com/skalogryz/fpCEF3

The bounty poster is currently sick, not sure if had a chance to contact with you.

**Comment by dliw** [comment]
Hi!
That's great news!

Do you have any plans how to proceed? I was planning to do an update to CEF 3.2785, but probably it would make sense to integrate your code changes first - if desired. The only problem I see, is that I don't have a Mac, so I cannot really maintain the Mac code - or test if things still work after an update.

>However, only Cocoa 64-bit is supported (due to CEF dropping 32-bit support for OSX).

I don't think this is a big restriction. The big majority of OSX installs today should be 64bit.

>The bounty poster is currently sick, not sure if had a chance to contact with you.

No, he did not contact me (yet).

**Comment by skalogryz** [comment]
> Do you have any plans how to proceed?

Me personally? No. My only interest is cash in this case (exchanged for the knowledge of pascal and osx).  I'll definitely provide support for the next month of this OSX port and fix any bugs, but no more than that. Naturally, I'll provide any assistance and comments on the changes made, while re-integrating into the main fpcef3 repository.

> The only problem I see, is that I don't have a Mac, so I cannot really maintain the Mac code

I strongly encourage you to contact the bounty poster (his email is shown in the bounty)! He has Mac, He is interested in keeping fpCef up-to-date.  He will learn OSX and Cocoa quickly. 
Any dramatic change to happen? I'm more than happy to assist with that for another bounty :) 

One thing I'd like to note, is that the latest Windows (32-bit) build is failing.
The reason for that, it doesn't like one of the "empty" strings but by default into configuration string. So I made a fix into the fork.

```
Index: cef3lib.pas
===================================================================
--- cef3lib.pas (revision 53)
+++ cef3lib.pas (revision 54)
@@ -344,8 +344,10 @@

 function CefString(const str: ustring): TCefString;
 begin
-  Result.str := PChar16(PWideChar(str));
   Result.length := Length(str);
+  if Result.length>0
+    then Result.str := PChar16(PWideChar(str))
+    else Result.str := nil;
   Result.dtor := nil;
 end;

```

**Comment by dliw** [comment]
Ok, I will try to integrate your fork ASAP. You could create a pull request, but I would prefer to clean up the code a little bit and create a single commit from it - if this is ok with you. Your contribution will be mentioned of course.

I have a two questions right now:

1. You did not change any code in the examples. Talking about *LCLSimple*: Does this work on Mac with your changes just as is? How needs the folder structure look like (app bundle) and where to put the subprocess? Like [this}(https://bitbucket.org/chromiumembedded/cef/wiki/Tutorial#markdown-header-mac-os-x-build-steps)?

It would help me a lot, if you could send me a working bundle (LCLSimple would be perfect). I think, I can then figure out, what is needed and extend the Readme accordingly.

2. You commented out [these lines](https://github.com/skalogryz/fpCEF3/commit/f2411f5094e1b3c6eb25ce368e7365ce3bbe2bd6#diff-38ca3cb3e62a93cc12ea26d026c76441). Are they only unnecessary or is there a problem? Could these callbacks be useful for users?

> One thing I'd like to note, is that the latest Windows (32-bit) build is failing.

In the commit you mentioned CEF 3.2704, but the latest fpCEF version is for CEF 3.2454. I didn't see any such problem, but the fix looks good.

**Comment by skalogryz** [comment]
> but I would prefer to clean up the code a little bit and create a single commit from it

I'm totally fine with any approach you'd like to take. I'm not a git savvy person, thus the less actions I need to take, the better for me.

> You did not change any code in the examples. Talking about LCLSimple: Does this work on Mac with your changes just as is?

Yes. The only adjustment has been made for LCLSimple is the path to subprocess. (an additional {$ifdef})
LCLSimple was also tested for work with a default subprocess path (an empty string) and as a single process model.
(Single process instructions are given here: https://github.com/skalogryz/fpCEF3/wiki )

> How needs the folder structure look like (app bundle) and where to put the subprocess

Subprocess goes into the main app bundle Contents/Frameworks/. It should be put into its own bundle and marked as an agent application.

I put instructions on this page: https://github.com/skalogryz/fpCEF3/wiki/Multiprocess-CEF-application

(see the note at the bottom about CEF naming conventions for OSX! ultimately "subprocess" shouldnot be used. Instead "<APPNAME> Helper" is the "expected" name by CEF )

> It would help me a lot, if you could send me a working bundle (LCLSimple would be perfect)

I can commit bundle structures to the repository (for LCLSimple and subprocess). But are you interested in having bundles with executables?

> You commented out these lines. Are they only unnecessary or is there a problem?

These methods are not implemented in the class (for an obvious reason), thus it would not compile for Cocoa. 
I simply commented them out, since they don't seem to be used in any way.
I presume they were ported here from a C++ example. Do you remember which one?

There's also a chance, that there's no need to implement these methods on fpCEF side, since they might be taken care of by LCL.

> Did you test off-screen-rendering?

I did not. Is there an example to be used for testing?

**Comment by dliw** [comment]
Thanks for your answers. And thanks for pointing me to your wiki, somehow I overlooked it. It answers my questions regarding how to create the app bundle. 

>I can commit bundle structures to the repository (for LCLSimple and subprocess).

Yes, please. This way I have a working example program and structure.

>But are you interested in having bundles with executables?

I cannot execute them and I will not add binaries into my branch. But you can commit them anyway, so I can see a fully working setup.

>I simply commented them out, since they don't seem to be used in any way. I presume they were ported here from a C++ example. Do you remember which one?

It's a straight conversion from the header file `include/capi/cef_browser_capi.h`. The pascal equivalent is [cef3api.pas](https://github.com/dliw/fpCEF3/blob/master/cef3api.pas#L631).
These are methods of the CEF browser host object to do some Mac specific actions. But they were removed in newer CEF versions, so no need to care.

>I did not. Is there an example to be used for testing?

It is the `OSRDemo`. It uses OpenGL, so probably this will work as well. At the moment only mouse handling is implemented - just in case you wonder why key input does not work.

**Comment by skalogryz** [comment]
> Yes, please. This way I have a working example program and structure.
There's a certain problem. svn bridge over git (used by Github) doesn't allow empty directories to be added.

I did commit whatever I could (excluding executables), but currently the bundle structure is incomplete in the repository.

I archived the bundle together and it should be available for your from this link
https://www.dropbox.com/s/kt5icce2947xh64/simple.zip?dl=0

**Comment by dliw** [comment]
>There's a certain problem. svn bridge over git (used by Github) doesn't allow empty directories to be added.

That's a limitation of git. No empty folders allowed.

>I archived the bundle together

Thank you very much.

**Comment by dliw** [comment]
@skalogryz
I just finished the integration in a new branch (https://github.com/dliw/fpCEF3/tree/macos-integration) and a wiki page (https://github.com/dliw/fpCEF3/wiki/macOS) with instructions.

Could you please test if everything works as expected?

**Comment by skalogryz** [comment]
you've an executable added as a part of the code repository:
/LCLSimple/simple.app/Contents/MacOS/simple

(I presume the same for subprocess) It should be removed!

**Comment by skalogryz** [comment]
LCLSimple with multiprocess works

**Comment by dliw** [comment]
Thank you very much for your quick feedback.

>you've an executable added as a part of the code repository
These should be empty files, so that git keeps the directory.

**Comment by skalogryz** [comment]
a few changes for subprocess project.
- setting a Cocoa widgetset as default (carbon would default and won't compile for 64 bit)

```
Index: subprocess.lpi
===================================================================
--- subprocess.lpi	(revision 56)
+++ subprocess.lpi	(working copy)
@@ -15,9 +15,6 @@
     <i18n>
       <EnableI18N LFM="False"/>
     </i18n>
-    <VersionInfo>
-      <StringTable ProductVersion=""/>
-    </VersionInfo>
     <BuildModes Count="4">
       <Item1 Name="Linux64" Default="True"/>
       <Item2 Name="Win32">
@@ -93,6 +90,9 @@
         </CompilerOptions>
       </Item3>
       <Item4 Name="macOS">
+        <MacroValues Count="1">
+          <Macro1 Name="LCLWidgetType" Value="cocoa"/>
+        </MacroValues>
         <CompilerOptions>
           <Version Value="11"/>
           <Target>
@@ -134,6 +134,9 @@
           </Other>
         </CompilerOptions>
       </Item4>
+      <SharedMatrixOptions Count="1">
+        <Item1 ID="266649750941" Modes="macOS" Type="IDEMacro" MacroName="LCLWidgetType" Value="cocoa"/>
+      </SharedMatrixOptions>
     </BuildModes>
     <PublishOptions>
       <Version Value="2"/>

```


```
Index: subprocess.lpr
===================================================================
--- subprocess.lpr	(revision 56)
+++ subprocess.lpr	(working copy)
@@ -3,9 +3,9 @@
 {$mode objfpc}{$H+}
 
 Uses
-  {$IFDEF UNIX}{$IFDEF UseCThreads}
+  {$IFDEF UNIX}
   cthreads,
-  {$ENDIF}{$ENDIF}
+  {$ENDIF}
   cef3types, cef3api;
 
 Var
```

**Comment by dliw** [comment]
Thanks, I integrated your changes.

Some more questions before I move the whole thing into master:

1. Does the `Execute after` command work as expected?
2. Is the stripped down version of your instructions in my wiki correct / understandable for Mac users?
3. Just to be sure: For the LCLSimple example: If I rename the subprocess to `simple Helper.app` and the subprocess executable to `simple Helper`, I can leave out `CefBrowserSubProcess` and everything works?

**Comment by skalogryz** [comment]
> 1. Does the Execute after command work as expected?

Yes it works as expected but only after I physically removed the file (that came from the repository). Prior to that it seemed like it would not overwrite it.

> Is the stripped down version of your instructions in my wiki correct / understandable for Mac users?

I cannot say for sure, if it's understandable :) Since I wrote ...pretty much for myself, in the first place. I'd put "multiprocess" version first. It's preferable and a recommended way by CEF anyway.

> Just to be sure: For the LCLSimple example: If I rename the subprocess to simple Helper.app and the subprocess executable to simple Helper, I can leave out CefBrowserSubProcess and everything works?

You also need to adjust Info.plist within (simple Helper.app). Info.plist points to the executable that must be run for the bundle. So it should also be changed from "subprocess" to "simple Helper"

```
Index: Info.plist
===================================================================
--- Info.plist	(revision 56)
+++ Info.plist	(working copy)
@@ -5,7 +5,7 @@
 	<key>CFBundleDevelopmentRegion</key>
 	<string>English</string>
 	<key>CFBundleExecutable</key>
-	<string>subprocess</string>
+	<string>simple Helper</string>
 	<key>CFBundleName</key>
 	<string>subprocess</string>
 	<key>CFBundleIdentifier</key>
```

**Comment by skalogryz** [comment]
OSRDemo works. 
(I'm not sure if it accepts keyboard input, but drawing via open gl and mouse events are working fine. I've been able to navigate on freepascal.org page)

And issue I ran into, is that CEF is being initialized prior to FormCreate call.
Thus assignment to CefBrowserSubprocessPath variable happens after the library initialization.

I moved the assigned to the initialization section of main.pas unit and it worked.

```
Index: main.pas
===================================================================
--- main.pas	(revision 56)
+++ main.pas	(working copy)
@@ -407,4 +407,11 @@
   fGLInitialized := True;
 end;
 
+initialization
+  CefBrowserSubprocessPath :=
+   ExtractFileDir(ExtractFileDir(ExpandFileName(Paramstr(0))))
+   +PathDelim+'Frameworks/subprocess.app/Contents/MacOS/subprocess';
+   ;
+
+
 end.
```

**Comment by dliw** [comment]
> Yes it works as expected but only after I physically removed the file

That's interesting. On Linux `cp` overwrites the target file unless it is a symbolic link. I will leave it to the Mac users to figure out how to fix this :smile:

> You also need to adjust Info.plist

Well ok, so I assume its better to just change the output name and create a new bundle

> OSRDemo works. I'm not sure if it accepts keyboard input
Thanks. Keyboard input is not implemented at all (on any platform).
I will add your the change and a **macOS** build mode.

One more thing:
The cef3 package should be installable as design time package even if you have the (default?) lazarus with Carbon interface. I don't know if there is a define to check whether a package is compiled for Lazarus or a user program, so I raise an exception in `TCustomChromium.Create` if the widget set is not Cocoa. I hope this exception does not behave too bad...

**Comment by skalogryz** [comment]
About the wiki and examples in general.

I'd recommend to mention that all projects are "Build mode" configured. (Not that many Lazarus users are using build modes... or are even aware of them).
For example, all examples are setup to compile for "Linux", by default. Whenever someone would attempt to compile an example project would run into an issue like "fcllaz" package not found. (Which makes no sense for fpcef3 package). 
This happens only because they didn't set the proper OS/CPU target. And they should do that by switching to the proper "Build mode" prepared for them. If users don't know that, they might adjust "linux" build-mode configuration instead
![screen shot 2016-12-27 at 11 50 50 am](https://cloud.githubusercontent.com/assets/16345231/21505146/b9a93ba2-cc32-11e6-90ae-2d228e2c8906.png)
.

**Comment by skalogryz** [comment]
> The cef3 package should be installable as design time package even if you have the (default?) lazarus with Carbon interface. I don't know if there is a define to check whether a package is compiled for Lazarus or a user program, so I raise an exception in TCustomChromium.Create if the widget set is not Cocoa. I hope this exception does not behave too bad...

Well, I can see it working for Lazarus Carbon in design item.

**Comment by dliw** [comment]
I pushed the changes for the OSRDemo.

> I'd recommend to mention that all projects are "Build mode" configured

Yes, I should have already done this. I extended the readme.

> For example, all examples are setup to compile for "Linux", by default

That's because Linux is the first build mode. The build mode is saved in the `*.lps` file, which for obvious reasons is not in git. Lazarus then uses the first mode as default.

**Comment by skalogryz** [comment]
> If you don't have any more suggestions for improvement, I will merge the branch into master.

Nothing from me at the moment. I'm sure OSX users feel provide more feedback soon

**Comment by dliw** [comment]
>I'm sure OSX users will provide more feedback soon

I hope so :smile:

Thank you very much again!

**Comment by skalogryz** [comment]
Lars Olson should be mentioned as the one who funded the contribution.

**Comment by dliw** [comment]
I changed the commit message (hopefully noone will be angry because of this push --force).
Is it ok to mention his name, it is nowhere on the bounty page?

**Comment by skalogryz** [comment]
I've already mentioned it. sorry. I cannot take all the credit by myself.

**Comment by dliw** [comment]
>I've already mentioned it. sorry.

As long as the founder is ok with it, there is no problem. I don't have contact with him.
